File: TitanicHaskell/logisticRegression.hs
module LogisticRegression (
	CostGrad,
	cost,
	grad,
	
	hypothesis,
	gradDescentOptimize,
	runRegression,
	predict
)
where

import Data.Matrix
import Control.Monad
import Debug.Trace

{-
	Definitions
	
	x,y - training set
	theta - parameters vector
	
	m - number of samples in training set
	n - number of features
		
	x     :: R[m*n]
	y     :: R[m*1]
	theta :: R[n*1]
-}

data CostGrad = CostGrad { cost :: Double, grad :: Matrix Double } deriving (Show)

sigmoid x = 1/(1 + exp(-x))

hypothesis :: Matrix Double -> Matrix Double -> Matrix Double
hypothesis theta x =
	fmap sigmoid $ x * theta

xlog x =
	if x == 0.0 then
		log 1e-17
	else
		log x

computeCostGrad :: Matrix Double -> Matrix Double -> Matrix Double -> Double -> CostGrad
computeCostGrad theta x y lambda =
	let
		m = fromIntegral $ nrows x
		n = fromIntegral $ ncols x
		sig = hypothesis theta x
		ones = matrix (nrows sig) (ncols sig) (\_ -> 1)
		
		theta1 = setElem 0 (1, 1) theta
		regVal = (lambda / (2.0 * m))*(sum $ elementwise (*) theta1 theta1)
		
		forOnes = (transpose $ fmap (* (-1)) y) * (fmap xlog sig)
		forZeroes = (transpose $ fmap (\z -> 1-z) y) * (fmap xlog $ fmap (\z -> 1-z) sig)
	in
		CostGrad {
			cost = ((forOnes - forZeroes) ! (1, 1))/m + regVal,
			grad = transpose (fmap (/ m) $ ((transpose (sig - y)) * x)) + (fmap (\i -> i*(lambda/m)) theta1)
		}

gradDescentOptimize :: Matrix Double -> Double -> (Matrix Double -> CostGrad) -> Double -> Integer -> Matrix Double
gradDescentOptimize theta alpha fx epsilon iterations =
	let
		c1 = fx theta
		newTheta = theta - (fmap (* alpha) $ grad c1)
		c2 = fx newTheta
		diff = ((cost c1 - cost c2) ** 2)
		diff1 = {-Debug.Trace.trace ("diff: " ++ show diff ++ " cost: " ++ (show $ cost c2) ++ " newTheta: " ++ show newTheta)$-} diff
	in
		if diff1 > epsilon && iterations > 0  then
			gradDescentOptimize newTheta alpha fx epsilon (iterations - 1)
		else
			newTheta

runRegression alpha lambda x y iterations =
	let
		theta = zero (ncols x) 1
	in
		gradDescentOptimize theta alpha (\z -> computeCostGrad z x y lambda) 0 iterations

predict theta x theshold =
	let
		estimatedProbabilities = hypothesis theta x
	in
		fmap (\x -> if x >= theshold then 1.0 else 0.0) estimatedProbabilities


testCost = do
	let x = fromLists [[1,2,3],[4,5,6],[7,8,9],[10,11,12]]
	let y = fromLists [[1],[0],[1],[0]]
	let theta = fmap (+1) $ zero 3 1
	let lambda = 1.5
	print "X:"
	print x
	print "Y:"
	print y
	print "Theta:"
	print theta
	print "Hypothesis"
	print $ hypothesis theta x
	let result = computeCostGrad theta x y lambda
	print "Cost:"
	print $ cost result
	print "Gradient:"
	print $ grad result
	let opt = gradDescentOptimize theta 0.01 (\z -> computeCostGrad z x y lambda) 0 3000
	print "Optimized theta:"
	print opt
	let newCost = cost $ computeCostGrad opt x y lambda
	print "New cost:"
	print newCost
	let rr = runRegression 0.01 lambda x y 1000
	print "runRegression:"
	print rr

File: TitanicPython/LogisticRegression.py
import numpy as np
import math as m
from scipy.optimize import minimize

vexp = np.vectorize(m.exp, otypes=[float])

def sigmoid(x):
    return 1/(1 + vexp(-x))

xlog = np.vectorize(lambda x: m.log(1e-17) if x == 0 else m.log(x))

def hypothesis(theta, x):
    return sigmoid(np.dot(x, theta))

def computeCostGrad(theta, x, y, lmb):
    (m, n) = x.shape
    thetaR = theta.reshape((n,1))
    sig = hypothesis(thetaR, x)
    theta1 = thetaR.copy()
    theta1[0] = 0
    regVal = (lmb/(2*m)) * (theta1*theta1).sum()
    forOnes = np.dot((-y).transpose(), xlog(sig))
    forZeroes = np.dot((1-y).transpose(), xlog(1 - sig))

    cost = ((forOnes - forZeroes).flat[0])/m + regVal
    grad = (np.dot((sig - y).transpose(), x)/m).transpose() + (lmb/m)*theta1

    return (cost, grad)

def costFun(theta, x, y, lmb):
    (cost, _) = computeCostGrad(theta, x, y, lmb)
    return cost

def gradFun(theta, x, y, lmb):
    (_, grad) = computeCostGrad(theta, x, y, lmb)
    return grad.reshape((grad.size,))

def runRegression(lmb, x, y):
    (nr, nc) = x.shape
    theta = np.zeros((nc, 1))
    thetaOpt = minimize(lambda p: costFun(p, x, y, lmb),
                    theta,
                    method='BFGS',
                    jac=lambda p: gradFun(p, x, y, lmb),
                    options={'disp': True})
    return thetaOpt.x.reshape(nc,1)

def squareError(x, y):
    return ((x-y)**2).sum()/(2*x.size)

def predictionError(x, y):
    return np.mean([1 if i[0] != i[1] else 0 for i in zip(x,y)])

def predict(theta, x, theshold):
    estimatedProbabilities = hypothesis(theta, x)
    return np.array([1.0 if p >= theshold else 0.0 for p in estimatedProbabilities])

# Do some tests
def doTests():
    x = np.array([[1,2,3],[4,5,6],[7,8,9],[10,11,12]], dtype=float)
    y = np.array([[1],[0],[1],[0]], dtype=float).reshape(4, 1)

    theta = np.ones((3,1))

    l = 1.5

    h = hypothesis(theta, x)
    print("Hypothesis: {0}".format(h))

    (cost, grad) = computeCostGrad(theta, x, y, l)
    print("Cost: {0}\nGrad: {1}".format(cost, grad))

    opt = gradDescentOptimize(theta, 0.01, lambda p: computeCostGrad(p, x, y, l), 3000)
    print("Optimized theta: {0}".format(opt))

    (newCost, _) = computeCostGrad(opt, x, y, l)
    print("New cost: {0}".format(newCost))

    rr = runRegression(0.01, l, x, y, 1000)

    print("runRegression: {0}".format(rr))

    pp = predict(rr, x, 0.5)
    print("Predicted results: {0}".format(pp));
